plugin/pkg/proxy/persistent.go

```diff
@@ -46,78 +47,49 @@ func newTransport(proxyName, addr string) *Transport {
 
 // connManager manages the persistent connection cache for UDP and TCP.
 func (t *Transport) connManager() {
-	ticker := time.NewTicker(defaultExpire)
-	defer ticker.Stop()
-Wait:
 	for {
 		select {
 		case proto := <-t.dial:
 			transtype := stringToTransportType(proto)
-			// take the last used conn - complexity O(1)
-			if stack := t.conns[transtype]; len(stack) > 0 {
-				pc := stack[len(stack)-1]
-				if time.Since(pc.used) < t.expire {
-					// Found one, remove from pool and return this conn.
-					t.conns[transtype] = stack[:len(stack)-1]
-					t.ret <- pc
-					continue Wait
-				}
-				// clear entire cache if the last conn is expired
-				t.conns[transtype] = nil
-				// now, the connections being passed to closeConns() are not reachable from
-				// transport methods anymore. So, it's safe to close them in a separate goroutine
-				go closeConns(stack)
-			}
-			t.ret <- nil
+			t.ret <- t.popValidConn(transtype)
 
 		case pc := <-t.yield:
 			transtype := t.transportTypeFromConn(pc)
 			t.conns[transtype] = append(t.conns[transtype], pc)
 
-		case <-ticker.C:
-			t.cleanup(false)
-
 		case <-t.stop:
-			t.cleanup(true)
+			for _, stack := range t.conns {
+				closeConns(stack)
+			}
 			close(t.ret)
 			return
 		}
 	}
 }
 
-// closeConns closes connections.
-func closeConns(conns []*persistConn) {
-	for _, pc := range conns {
-		pc.c.Close()
-	}
-}
-
-// cleanup removes connections from cache.
-func (t *Transport) cleanup(all bool) {
-	staleTime := time.Now().Add(-t.expire)
-	for transtype, stack := range t.conns {
-		if len(stack) == 0 {
-			continue
-		}
-		if all {
-			t.conns[transtype] = nil
-			// now, the connections being passed to closeConns() are not reachable from
-			// transport methods anymore. So, it's safe to close them in a separate goroutine
-			go closeConns(stack)
+func (t *Transport) popValidConn(transtype transportType) *persistConn {
+	now := time.Now()
+	for idx, pc := range t.conns[transtype] {
```

Comment:
This becomes O(n) instead of the previous O(1) fetching of a connection. On normal operations, I would expect most connections to be valid and only needing to iterate through a few connections before finding a valid one. Benchmarks indicate that each step of the loop is about 5ns.